File: apps/web/src/app/architecture/catalog.rs
//! Stable architecture levels, operations, and config-driven node catalog.

use nanogpt_schema::GptConfig;

/// Stable hierarchy levels exposed by the architecture map.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub(crate) enum ArchitectureLevel {
    /// Full GPT forward path.
    #[default]
    Gpt,
    /// One selected Transformer block.
    Block,
    /// One selected block's multi-head attention.
    Attention,
    /// Autoregressive generation loop.
    Generation,
}

impl ArchitectureLevel {
    /// Exact public hierarchy catalog.
    #[cfg(test)]
    pub(crate) const ALL: [Self; 4] = [Self::Gpt, Self::Block, Self::Attention, Self::Generation];

    /// Stable machine-readable level key.
    #[must_use]
    pub(crate) const fn slug(self) -> &'static str {
        match self {
            Self::Gpt => "gpt",
            Self::Block => "block",
            Self::Attention => "attention",
            Self::Generation => "generation",
        }
    }
}

/// Selectable computation boundaries in the architecture map.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum ArchitectureOperation {
    Embedding,
    FinalLayerNorm,
    LanguageModelHead,
    AttentionLayerNorm,
    AttentionResidual,
    MlpLayerNorm,
    Mlp,
    MlpResidual,
    Query,
    Key,
    Value,
    QueryKeyProduct,
    Scale,
    Mask,
    Softmax,
    ValueProduct,
    MergeHeads,
    Projection,
    Logits,
    Temperature,
    TopK,
    GenerationSoftmax,
    Sample,
    Append,
    Repeat,
}

/// Summary tensor identity selected by an architecture operation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum SummaryEvidence {
    /// Final normalized residual stream.
    FinalLayerNorm,
    /// Vocabulary logits from the tied language-model head.
    Logits,
}

/// One selectable node kind; labels remain presentation-only.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum ArchitectureNodeKind {
    Operation(ArchitectureOperation),
    Layer(usize),
    Head(usize),
    Level(ArchitectureLevel),
}

/// Catalog entry used by the Leptos component.
#[cfg(any(test, target_arch = "wasm32"))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct ArchitectureNode {
    pub(crate) kind: ArchitectureNodeKind,
}

/// Builds the exact level catalog from model configuration.
#[cfg(any(test, target_arch = "wasm32"))]
#[must_use]
pub(crate) fn catalog(config: &GptConfig, level: ArchitectureLevel) -> Vec<ArchitectureNode> {
    use ArchitectureLevel as L;
    use ArchitectureNodeKind::{Head, Layer, Level, Operation};
    use ArchitectureOperation as O;
    let kinds = match level {
        L::Gpt => std::iter::once(Operation(O::Embedding))
            .chain((0..config.n_layer).map(Layer))
            .chain([
                Operation(O::FinalLayerNorm),
                Operation(O::LanguageModelHead),
                Level(L::Generation),
            ])
            .collect(),
        L::Block => vec![
            Operation(O::AttentionLayerNorm),
            Level(L::Attention),
            Operation(O::AttentionResidual),
            Operation(O::MlpLayerNorm),
            Operation(O::Mlp),
            Operation(O::MlpResidual),
        ],
        L::Attention => (0..config.n_head)
            .map(Head)
            .chain(
                [
                    O::Query,
                    O::Key,
                    O::Value,
                    O::QueryKeyProduct,
                    O::Scale,
                    O::Mask,
                    O::Softmax,
                    O::ValueProduct,
                    O::MergeHeads,
                    O::Projection,
                ]
                .into_iter()
                .map(Operation),
            )
            .collect(),
        L::Generation => [
            O::Logits,
            O::Temperature,
            O::TopK,
            O::GenerationSoftmax,
            O::Sample,
            O::Append,
            O::Repeat,
        ]
        .into_iter()
        .map(Operation)
        .collect(),
    };
    kinds
        .into_iter()
        .map(|kind| ArchitectureNode { kind })
        .collect()
}

File: apps/web/src/architecture/catalog.ts
export type ArchitectureNodeId =
  | "root"
  | "input-context"
  | "token-embedding"
  | "position-embedding"
  | "embedding-add"
  | "hidden-state"
  | "transformer-block"
  | "layer-norm-1"
  | "self-attention"
  | "residual-1"
  | "layer-norm-2"
  | "mlp"
  | "residual-2"
  | "final-layer-norm"
  | "lm-head"
  | "logits"
  | "token-selection"
  | "generated-token"
  | "append-context"
  | "attention-qkv-projection"
  | "attention-query"
  | "attention-key"
  | "attention-value"
  | "attention-scores"
  | "attention-scale"
  | "attention-causal-mask"
  | "attention-softmax"
  | "attention-value-aggregation"
  | "attention-merge-heads"
  | "attention-output-projection";

export type ArchitectureNodeCapability = "static" | "selectable" | "drill-down";

export interface ArchitectureNodeDefinition {
  readonly id: ArchitectureNodeId;
  readonly capability: ArchitectureNodeCapability;
  readonly accessibleName: string;
}

export const ARCHITECTURE_NODE_IDS: readonly ArchitectureNodeId[] = [
  "root",
  "input-context",
  "token-embedding",
  "position-embedding",
  "embedding-add",
  "hidden-state",
  "transformer-block",
  "layer-norm-1",
  "self-attention",
  "residual-1",
  "layer-norm-2",
  "mlp",
  "residual-2",
  "final-layer-norm",
  "lm-head",
  "logits",
  "token-selection",
  "generated-token",
  "append-context",
  "attention-qkv-projection",
  "attention-query",
  "attention-key",
  "attention-value",
  "attention-scores",
  "attention-scale",
  "attention-causal-mask",
  "attention-softmax",
  "attention-value-aggregation",
  "attention-merge-heads",
  "attention-output-projection",
];

function node(
  id: ArchitectureNodeId,
  capability: ArchitectureNodeCapability,
  accessibleName: string,
): ArchitectureNodeDefinition {
  return { id, capability, accessibleName };
}

export const architectureNodeCatalog: Readonly<
  Record<ArchitectureNodeId, ArchitectureNodeDefinition>
> = {
  root: node("root", "static", "GPT Architecture"),
  "input-context": node(
    "input-context",
    "selectable",
    "현재 input token sequence",
  ),
  "token-embedding": node(
    "token-embedding",
    "selectable",
    "Token embedding lookup",
  ),
  "position-embedding": node(
    "position-embedding",
    "selectable",
    "Position embedding lookup",
  ),
  "embedding-add": node(
    "embedding-add",
    "static",
    "Token과 Position embedding의 원소별 덧셈",
  ),
  "hidden-state": node("hidden-state", "static", "초기 hidden state X zero"),
  "transformer-block": node(
    "transformer-block",
    "drill-down",
    "반복 Transformer Blocks",
  ),
  "layer-norm-1": node(
    "layer-norm-1",
    "selectable",
    "LayerNorm 1, Block 입력 정규화",
  ),
  "self-attention": node(
    "self-attention",
    "drill-down",
    "Causal Multi-Head Self-Attention",
  ),
  "residual-1": node("residual-1", "selectable", "첫 번째 residual 덧셈"),
  "layer-norm-2": node(
    "layer-norm-2",
    "selectable",
    "LayerNorm 2, 첫 residual 결과 정규화",
  ),
  mlp: node("mlp", "selectable", "MLP feed-forward network"),
  "residual-2": node("residual-2", "selectable", "두 번째 residual 덧셈"),
  "final-layer-norm": node("final-layer-norm", "selectable", "Final LayerNorm"),
  "lm-head": node("lm-head", "selectable", "LM Head Linear projection"),
  logits: node("logits", "selectable", "Vocabulary logits"),
  "token-selection": node("token-selection", "selectable", "다음 token 선택"),
  "generated-token": node("generated-token", "selectable", "생성된 token"),
  "append-context": node(
    "append-context",
    "selectable",
    "생성 token을 context에 추가",
  ),
  "attention-qkv-projection": node(
    "attention-qkv-projection",
    "selectable",
    "QKV Projection, 하나의 Linear layer",
  ),
  "attention-query": node("attention-query", "selectable", "Query tensor Q"),
  "attention-key": node("attention-key", "selectable", "Key tensor K"),
  "attention-value": node("attention-value", "selectable", "Value tensor V"),
  "attention-scores": node(
    "attention-scores",
    "selectable",
    "Score MatMul, Query와 전치된 Key의 행렬곱",
  ),
  "attention-scale": node(
    "attention-scale",
    "selectable",
    "Scale, score를 head dimension 제곱근으로 나누기",
  ),
  "attention-causal-mask": node(
    "attention-causal-mask",
    "selectable",
    "Causal Mask, 미래 token position 차단",
  ),
  "attention-softmax": node(
    "attention-softmax",
    "selectable",
    "Softmax, attention probability 정규화",
  ),
  "attention-value-aggregation": node(
    "attention-value-aggregation",
    "selectable",
    "Value MatMul, attention probability와 Value의 행렬곱",
  ),
  "attention-merge-heads": node(
    "attention-merge-heads",
    "selectable",
    "Merge Heads, head output 연결과 reshape",
  ),
  "attention-output-projection": node(
    "attention-output-projection",
    "selectable",
    "Output Projection, Linear C to C",
  ),
};
